### PR TITLE
feat: wasi 

### PR DESCRIPTION
**Pull request by Mesteery**
Depends on: #37 

*Tested (Deno & browser) and works well.*

### CODE REVIEW

File: src-browser/wasi.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+export class WASI extends _WASI {
+  constructor(options) {
+    super({
+      ...options,
+      bindings: {
+        ..._WASI.defaultBindings,
+        isTTY: () => false,
+        fs,
+      },
+    });
```
Comment (guybedford) [inline src-browser/wasi.js:13]:
```suggestion
    super(Object.assign({}, options, {
      bindings: Object.assign({}, _WASI.defaultBindings, {
        isTTY: () => false,
        fs,
      }),
    }));
```

**Review by guybedford — COMMENTED** [review]
This is really incredible to see! I didn't know about the Wasmer shim, exciting stuff.

Because WASI is prety much an entire VM that we're bootstrapping here, and since there is custom stuff going on that brings together various functionality - I think it's super important to have an end to end WASI integration test that shows the VM can execute a simple application. Otherwise we seriously risk breaking this in future changes.

Perhaps there are some simple subsets of the Node.js cases to be taken from https://github.com/nodejs/node/tree/master/test/wasi.

Alternatively something super lazy like a Wasmer cowsay WASI build demonstrating output or another similarly simple to very WASI execution. I'm not expecting every stdlib call to be tested - just something that shows it bootstraps and produces output is fine.

**Comment by Mesteery** [comment]
(I'm adding tests, but I'm not done yet.)

**Comment by guybedford** [comment]
I've pushed up a WASI test case here for you.

It's the following simple case that works in Node.js via `node --experimental-wasi-unstable-preview1 wasi.mjs`:

```js
import { readFile } from 'fs/promises';
import { WASI } from 'wasi';

const wasi = new WASI({
  args: ['cowsay', 'moo']
});

const importObject = { wasi_unstable: wasi.wasiImport };

const wasm = await WebAssembly.compile(await readFile(new URL('./cowsay.wasm', import.meta.url)));
const instance = await WebAssembly.instantiate(wasm, importObject);

wasi.start(instance);
```

Currently I'm hitting the issue that the `readFile` implementation is unable to read from external URLs. That might be a nice feature to add to provide a universal subset here - otherwise we would need to do some dancing around working what paths to use in the virtual Node.js filesystem, when ideally we should be focused on the URLs as the universal subset I think.

**Comment by guybedford** [comment]
The memory FS is saying that localhost is required for all file loads... I'm not sure that will work for us unfortunately...??

Some options:
1. Set up FS to work against URL objects for async fs methods where suitable. It's a hack, but it gets somewhere and supprots at least a basic readFile.
2. Extend MemFS to support arbitrary URLs using URL object inputs, where all network operations are cached into the memFS and sync operations are permitted into the cache. This would allow "priming" the memfs via eg `await readFile(new URL('https://external.com/file.js'))` then permitting `readFileSync` to work. Because we own the filesystem and URLs in this context can be considered immutable, we don't need to worry about unknown writes, since all other writes must go through memfs anyway.

(2) sounds like a nice generic solution here. (1) as an initial step will get WASI working in the browser.

**Comment by Mesteery** [comment]
> The memory FS is saying that localhost is required for all file loads...

I did not understand

**Comment by guybedford** [comment]
@Mesteery try running the test I just pushed up to this PR and you will see the error.

**Comment by Mesteery** [comment]
Okay, I see. I am for the second solution. But I think we should maintain a fork of memfs (which seems to be unmaintained).

**Comment by guybedford** [comment]
> But I think we should maintain a fork of memfs (which seems to be unmaintained).

This sounds sensible to me. It would be beneficial to treat it as its own package / project too.

**Comment by Mesteery** [comment]
How to do otherwise?

**Comment by guybedford** [comment]
An alternative might be to try and hack a wrapper around memfs temporarily for now.

**Comment by Mesteery** [comment]
Okay. By the way, sorry, I confused the 'sensible' with 'sensitive'. (that's why I sent `How to do otherwise?`) 😅

**Comment by guybedford** [comment]
Ah ok! It would be great to see an implementation something like:

* MemFS, but make the pathing scheme URLs, where `/asdf` means `file:///asdf`.
* For network URLs, `readFile` delegates to `fetch` if not found in the memfs, which then populates the memfs, using the memfs otherwise.

It's very much a network wrapper around MemFS still, but as its own project can be tested and used more widely.

**Comment by Mesteery** [comment]
Yes, that's what I was thinking: for each read/stat or similar method, the file is searched in the memory, if not found, in the network.
But there are some questions :
- What if the host is not the same as on the current page?
- How to manage the cwd of memfs, and the base of the URL (this would break things, if for example the cwd is changed and would create inconsistencies).
- If it's a folder that doesn't exist in memory, should we look for it in the network?

Translated with www.DeepL.com/Translator (free version)

**Comment by guybedford** [comment]
`file:///` forms a virtual file system scheme, which unless otherwise populated in the same session is a completely empty filesystem.

* A _path_ like `/asdf` is a file system path and hence is always `file:///asdf`.
* A _url object_ input to a file method is treated as a URL only. Eg `new URL('https://site.com/file.js')`. These effectively become other host-based file system paths.
* `process.cwd()` currently returns `/` in the browser I believe? `path.resolve('asdf')` should also return `/asdf` in the browser as well I believe. That does confuse the `https://host.com/` and `file:///` endpoint locations though agreed. As a string, strictly speaking, `/asdf` shoudl always be the memfs `file:///asdf` so the `cwd` as the base of the memfs.
* This could be configurable via setting the cwd, but then we don't support setting the cwd to a URL. But I think that matches the model in that all paths must be absolute file-style paths agains the file scheme.

**Comment by guybedford** [comment]
See https://nodejs.org/dist/latest-v16.x/docs/api/fs.html#fs_file_url_paths for info on URL paths in fs methods.

**Comment by Mesteery** [comment]
I was actually thinking of supporting chdir, but when I think about it it's just a bad idea. So yes the cwd is always '/'.

If I understand correctly :
```js
fs.readFileSync('./test.wasm'); // error

// assuming that test.wasm exists in the network
fs.readFileSync(new URL('./test.wasm', import.meta.url)); // file contents

fs.readFileSync(new URL('./test.wasm', import.meta.url)); // fetched from the memory (cached)

fs.readFileSync('./test.wasm'); // error

// assuming that test.wasm exists
fs.readFileSync(new URL('https://another-host.com/test.wasm')); // file contents
fs.readFileSync('https://another-host.com/test.wasm'); // error
```

**Comment by guybedford** [comment]
Here's what I'm imagining here, let me know if this makes sense to you:

```js
// Resolves to /test.wasm (CWD), checks file:///test.wasm in memfs, throws when not found
fs.readFileSync('./test.wasm');

// Works out, because we are writing and then reading the existing file in memfs:
fs.writeFileSync('/test.wasm', 'test');
fs.readFileSync('./test.wasm').toString() === 'test';

// Resolves to https://baseurl.com/test.wasm, checks memfs, throws if not found with error "network URLs must be fetched async unless already loaded"
fs.readFileSync(new URL('./test.wasm', import.meta.url)); 

// Async FS calls can wait on network:
// Underneath it just does:
//   fs.writeFileSync('https://baseurl.com/test.wasm', await (await fetch(url)).arrayBuffer())
//   return fs.promises.readFile('https://baseurl.com/test.wasm');
await fs.promises.readFile(new URL('./test.wasm', import.meta.url));

// Works out, because we've already loaded above so it is in the memfs
fs.readFileSync(new URL('./test.wasm', import.meta.url);
```

**Comment by Mesteery** [comment]
It makes sense to me.
However, I have a few questions:
- is an http url a valid path? Or only if it is a URL object?
- Is it possible to do `await fsPromises.readFile('./network-file' /* or /network-file, ... */)`? (without URL).

**Comment by guybedford** [comment]
> is an http url a valid path? Or only if it is a URL object?

Only if it is a URL object. Whether there is a string representation that matches is a good question. Eg we could support `//url/https://site.com/...` as a special type of "URL Path" so that the scheme is actually a path scheme embedding a URL scheme. Then file URLs would always be of the path form converting `file:///some/path` into `/some/path` and not using `//url/file:///some/path`.

> `await fsPromises.readFile('./network-file' /* or /network-file, ... */)`

This would then effectively be `await fsPromises.readFile('//url/https://base.com/network-file')`.

**Comment by Mesteery** [comment]
Yes, so unfortunately we can't be compatible with Node.js on this point.
But this is an interesting feature as it is a project that will not necessarily be limited to JSPM and Node.js-browser compatibility.
So I think I understood.

**Comment by guybedford** [comment]
@Mesteery I've got WASI running in the browser using the approach described above. Please take a look and let me know what you think.

File: src-browser/fs.js [hunk 2]
```diff
@@ -60,6 +61,34 @@ fs.StatWatcher.prototype.start = function (path, persistent, interval = 5007) {
 fs.FileReadStream = fs.ReadStream;
 fs.FileWriteStream = fs.WriteStream;
 
+function handleFsUrl (url, isAsync) {
+  if (url.protocol === 'file:')
+    return fileURLToPath(url);
+  if (url.protocol === 'https:' || url.protocol === 'http:') {
+    const path = '\\\\url\\' + url.replaceAll(/\//g, '\\\\');
```
Comment (Mesteery) [inline src-browser/fs.js:68]:
```suggestion
    const path = '\\\\url\\' + url.href.replaceAll(/\//g, '\\\\');
```

**Comment by Mesteery** [comment]
Personally, I think it is interesting as a temporary solution. I would still like to know how memfs behaves with such a path (`\\url\...`).

**Pull request merged**